lib/ASTGen/Sources/ASTGen/Exprs.swift

```diff
@@ -1,6 +1,65 @@
 import ASTBridging
 import SwiftDiagnostics
-import SwiftSyntax
+@_spi(ExperimentalLanguageFeatures) import SwiftSyntax
+
+/// Check if an `ExprSyntax` can be generated using ASTGen.
+///
+/// If all the expression nodes that shares the first token are migrated,
+/// returns true. For example, given
+///   ```
+///   foo.bar({ $0 + 1 }) + 2
+///   ```
+/// `foo` token is the first token of all `SequenceExpr`, `FunctionCallExpr`,
+/// `MemberAccessExpr`, and `DeclReferenceExpr`. All these expression kinds must
+/// be migrated to handle it in ASTGen. Because the fallback
+/// `generateWithLegacy(_:)` only receives the parser position, it eagerly
+/// parses the outer expressions.
+func isExprMigrated(_ node: ExprSyntax) -> Bool {
+  var current: Syntax = Syntax(node)
+  if let firstToken = node.firstToken(viewMode: .sourceAccurate) {
+    current = firstToken.parent!
+  }
+  while true {
+    switch current.kind {
+    case // Known implemented kinds.
+        .closureExpr, .functionCallExpr, .declReferenceExpr, .memberAccessExpr,
+        .tupleExpr, .ifExpr:
+
+      // NOTE: When SequenceExpr is implemented, we need some special handling.
+      // E.g.
+      //   <implemented> + <unimplemented> + <implemented>
+      // If we delegate `<unimplemented>` part to the legacy parser, it would
+      // eagerly parse the rest of the expression, instead of just the
+      // <unimplemented> part.
+      // Maybe call 'Parser::parseExprSequenceElement' directly.
+      break
+    case // Known unimplemented kinds.
+        .arrayExpr, .arrowExpr, .asExpr, .assignmentExpr, .awaitExpr,
+        .binaryOperatorExpr, .booleanLiteralExpr, .borrowExpr, .canImportExpr,
```

Comment:
Literals are already implemented. I didn't notice they were in `Literals.swift`.